### ISSUE TITLE
[Feature request] unsquashing unsorted_segment_x

### ISSUE THREAD

**Issue by roya0045** [tensorflow/tensorflow#18311]
Have I written custom code? No
OS Platform and Distribution? Win10
TensorFlow installed from? pip
TensorFlow version? 1.7
Bazel version? N/A
CUDA/cuDNN version? N/A
GPU model and memory? N/A
Exact command to reproduce: unsorted_segment_sum

### Describe the problem
Let's say you have 3 sequences of 12 values (shape (3,12)) with 5 segments and you have an input of shape (None,12,3) transposed to (3,12,None)
using any function with unsorted_segement_x the output will be (5,None).

Would it be possible to have a function that does the same but still keep the first dimention.
Something equivalent to:
```python
input.shape==(None,12,3)
segment.shape==(3,12)

outputs=[]
for ix, val in enumerate(tf.unstack(input)):
    ouputs.append(tf.unsorted_segment_sum(val,sequence[ix],5)
output=tf.stack(outputs)
output.shape==(3,5,None)
```
The use of such a thing would be to iterate over a weighted input and without having to loop over the data or using reshape. I think this should be somewhat straightforward to implement; not concatenate together all the sequences for the superior dimensions and use "keep_dims" similar to reduce_sum to trigger that behaviour.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by roya0045** [comment]
Have I written custom code? No
OS Platform and Distribution? Win10
TensorFlow installed from? pip
TensorFlow version? 1.7
Bazel version? N/A
CUDA/cuDNN version? N/A
GPU model and memory? N/A
Exact command to reproduce: unsorted_segment_sum
(duplicate info for the bot)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by zhaoxin19** [comment]
I want to use the same feature too! It is strange that the segment_ops don't have keep_dims parameter.

**Comment by roya0045** [comment]
@zhaoxin19 There are some workaround to this, a good idea might be to start a new thread about this since @rmlarsen  seems to be deceased. If you are in a rush and need some workarounds I could help.

**Comment by zhaoxin19** [comment]
@roya0045 ,thank you very much for you kind help. I am doing deep model research and want to add this kind of operation in my new model. Lucily I've found the unsorted_segment_sum implement. But it is a little different as what I want. I think add a keep_dims parameter is a good idea in this operation. I'm not expert in adding or changing a new operation in tensorflow. I will be very grateful if anyone could help.

**Comment by zhaoxin19** [comment]
@roya0045 do you know how can we add this feature in tensorflow? Seems that we just need to change tf.unsorted_segment_sum api as tf.reduce_sum adding a keep_dims parameter.

**Comment by roya0045** [comment]
I'm sure it can be done but I'm not sure how. I'm not familiar with c++ and the code base for tensorflow.

The workaround I have found so far are implemented in [ my latest projet](https://github.com/roya0045/Dendritic_layer-TF-Keras-/blob/02cf170356fb0109bbcb565786cb656085be7245/tf_dendritic.py#L237)

The only issue is that I have not tested if those 4 methods break backpropagation (get the weights first, train on random data, get the weight after and just subtract the initial weight to see if it worked.)

I wanted to test them but I haven't had much time to do so, if you want you could just set up a minimal network with a for loop for each version , train it and compare the weights to see if those workarounds are valid, I'd be interested in the results.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 48 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 64 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 80 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 94 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 109 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 124 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 140 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 155 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
